Make possible for apps to define their own avatar types

danxuliu opened the pull request:
Requires #24578

I am really sorry to come with this change so late in the cycle. It is perfectly fine if it is deferred to Nextcloud 22 :-) (or if it is a bad idea and it is never merged ;-) ). However, while working on the [_Group conversation picture_ feature for Talk](https://github.com/nextcloud/spreed/issues/927) I thought that it would be good to generalize the avatars and use them for that. This would make possible, for example, to show the avatar of a room in the list of shares without needing to use a Talk endpoint.

I guess it could be used too for [unifying the avatars between clients](https://github.com/nextcloud/server/issues/23467) or [differentiating user and group avatars](https://github.com/nextcloud/server/issues/12780) (although that would not be something to be done for Nextcloud 21).

Before this pull request the avatars were limited to users and guests, and there was no OCS endpoint for getting and setting the avatars. Now, besides the default user and guest types, an app can register its own avatar type(s), and any avatar can be got or set using a OCS endpoint (and the old ones are still available, of course).

In order to set an avatar with the old endpoints a temporary image needs to be posted, and then a request to crop that temporary image is needed to finally set the avatar. When using the OCS endpoint, on the other hand, the image needs to be uploaded already with the desired size (and it can not be selected from the internal files, the client setting the avatar needs to download it first, crop it as needed and then upload it again). If the avatar does not support the aspect ratio (for example, the user avatar must be squared) then the request fails.

Pending:
- [X] Add `IAvatarProvider` (or something like that) and registration of providers to `IAvatarManager`
- [X] Prevent user avatars from being modified by other users
  - Should it be done in `UserAvatar`? Or should be prevented only in the controller (for example, to allow an admin to modify the avatar from an OCC command, or something like that)? In case it needs to be done in the controller, how? Add something like `canAvatarBeModifiedByUser($userId)` to `IAvatar` and check it from the controller?
- [X] Use different cache values in the controller depending on the avatar type
  - Where to specify the cache value to use for each type? With "getCache()" in the AvatarProvider? With a more generic "getAvatarOptions()" (or something similar) that returns a map of keys and values so IAvatarProvider does not need to be extended? In the IAvatar itself so it could even change depending on the avatar instance?
- [X] Add a method to get the avatar version to IAvatarProvider, similar to `getCacheTimeToLive(IAvatar)`

Future:
- Create a `UserAvatarProvider` and move [code specific to user avatars](https://github.com/nextcloud/server/blob/cb057829f72c70e819f456edfadbb29d72dba832/lib/private/Avatar/AvatarManager.php#L98-L144) from `AvatarManager` to `UserAvatarProvider` (not done yet to minimize the changes). - `UserAvatarProvider` has been created already, as getting the cache will be done from `IAvatarProvider`, although the code specific to user avatars was not moved yet to the provider.

danxuliu committed:
Make possible to send requests as anonymous users in integration tests

Until now requests always had "auth" headers either for an admin or a
regular user, depending on the value of "currentUser". Now, if
"currentUser" starts by "anonymous" no "auth" header is sent, which
makes possible to also test requests with users not logged in.

Signed-off-by: Daniel Calviño Sánchez <danxuliu@gmail.com>

danxuliu committed:
Make possible to set body in requesttoken requests in integration tests

"sendingAToWithRequesttoken" needs to be used to test some non OCS
endpoints which require the request token to be sent in the request. Now
it is possible to specify the body (or, rather, additional contents
beside the cookies and the request token) for those requests, as it will
be needed for example to upload an avatar.

Signed-off-by: Daniel Calviño Sánchez <danxuliu@gmail.com>

danxuliu committed:
Add integration tests for user avatars

Signed-off-by: Daniel Calviño Sánchez <danxuliu@gmail.com>

danxuliu committed:
Add integration tests for resized user avatars

Even on solid color images the resizing can cause some small artifacts
that slightly modify the color of certain pixels. Due to this now the
color comparison is no longer strict but fuzzy.

Signed-off-by: Daniel Calviño Sánchez <danxuliu@gmail.com>

danxuliu committed:
Add integration tests for getting guest avatars

Signed-off-by: Daniel Calviño Sánchez <danxuliu@gmail.com>

danxuliu committed:
Add getter for generic avatars to IAvatarManager

Signed-off-by: Daniel Calviño Sánchez <danxuliu@gmail.com>

danxuliu committed:
Add OCS endpoint for avatars

Signed-off-by: Daniel Calviño Sánchez <danxuliu@gmail.com>

danxuliu committed:
Make possible to set body in OCS requests in integration tests

"sendingToWith" is used to test OCS endpoints. Now it is possible to
specify the body (or, rather, additional contents beside the cookies and
the request token) for those requests, as it will be needed for example
to upload an avatar.

Signed-off-by: Daniel Calviño Sánchez <danxuliu@gmail.com>

danxuliu committed:
Add integration tests for getting and setting generic avatars

Signed-off-by: Daniel Calviño Sánchez <danxuliu@gmail.com>

nickvergessen commented on lib/private/Avatar/AvatarManager.php:
```diff
@@ -152,4 +152,16 @@ public function deleteUserAvatar(string $userId): void {
+			return $this->getGuestAvatar($id);
+		}
+
+		throw new \InvalidArgumentException('Unknown avatar type: ' . $type);
```
How can apps specify/allow their types? I don't see an event or method to allow it?

danxuliu replied:
It was not implemented yet. Now it is done through the `IRegistrationContext` (but please verify that this is the right way to do it).

nickvergessen commented on core/Controller/GenericAvatarController.php:
```diff
@@ -0,0 +1,173 @@
+	 *
+	 * @return JSONResponse
+	 */
+	public function deleteAvatar(string $avatarType, string $avatarId) {
```
```suggestion
	public function deleteAvatar(string $avatarType, string $avatarId): DataResponse {
```

nickvergessen commented on core/Controller/GenericAvatarController.php:
```diff
@@ -0,0 +1,173 @@
+	 * @param string $path
+	 * @return JSONResponse
+	 */
+	public function setAvatar(string $avatarType, string $avatarId) {
```
```suggestion
	public function setAvatar(string $avatarType, string $avatarId): DataResponse {
```

nickvergessen commented on core/Controller/GenericAvatarController.php:
```diff
@@ -0,0 +1,173 @@
+	 * @PublicPage
+	 *
+	 * @param string $path
+	 * @return JSONResponse
```
```suggestion
	 * @return DataResponse
```

nickvergessen commented on core/Controller/GenericAvatarController.php:
```diff
@@ -0,0 +1,173 @@
+				]
+			);
+		} catch (\Exception $e) {
+			return new JSONResponse([], Http::STATUS_NOT_FOUND);
```
```suggestion
			return new DataResponse([], Http::STATUS_NOT_FOUND);
```

nickvergessen commented on core/Controller/GenericAvatarController.php:
```diff
@@ -0,0 +1,173 @@
+	 * @param int $size
+	 * @return JSONResponse|FileDisplayResponse
+	 */
+	public function getAvatar(string $avatarType, string $avatarId, int $size) {
```
```suggestion
	public function getAvatar(string $avatarType, string $avatarId, int $size): Response {
```

nickvergessen commented on core/Controller/GenericAvatarController.php:
```diff
@@ -0,0 +1,173 @@
+	 * @param string $avatarType
+	 * @param string $avatarId
+	 * @param int $size
+	 * @return JSONResponse|FileDisplayResponse
```
```suggestion
	 * @return DataResponse|FileDisplayResponse
```

nickvergessen commented on core/Controller/GenericAvatarController.php:
```diff
@@ -0,0 +1,173 @@
+use OCP\AppFramework\OCSController;
+use OCP\AppFramework\Http;
+use OCP\AppFramework\Http\FileDisplayResponse;
+use OCP\AppFramework\Http\JSONResponse;
```
```suggestion
use OCP\AppFramework\Http\DataResponse;
```

nickvergessen commented on core/Controller/GenericAvatarController.php:
```diff
@@ -0,0 +1,173 @@
+		$files = $this->request->getUploadedFile('files');
+
+		if (is_null($files)) {
+			return new JSONResponse(
```
```suggestion
			return new DataResponse(
```

nickvergessen commented on core/Controller/GenericAvatarController.php:
```diff
@@ -0,0 +1,173 @@
+			!is_uploaded_file($files['tmp_name'][0]) ||
+			\OC\Files\Filesystem::isFileBlacklisted($files['tmp_name'][0])
+		) {
+			return new JSONResponse(
```
```suggestion
			return new DataResponse(
```

nickvergessen commented on core/Controller/GenericAvatarController.php:
```diff
@@ -0,0 +1,173 @@
+		}
+
+		if ($files['size'][0] > 20 * 1024 * 1024) {
+			return new JSONResponse(
```
```suggestion
			return new DataResponse(
```

nickvergessen commented on core/Controller/GenericAvatarController.php:
```diff
@@ -0,0 +1,173 @@
+		try {
+			$avatar = $this->avatarManager->getGenericAvatar($avatarType, $avatarId);
+			$avatar->set($image);
+			return new JSONResponse(
```
```suggestion
			return new DataResponse(
```

nickvergessen commented on core/Controller/GenericAvatarController.php:
```diff
@@ -0,0 +1,173 @@
+				['status' => 'success']
+			);
+		} catch (\OC\NotSquareException $e) {
+			return new JSONResponse(
```
```suggestion
			return new DataResponse(
```

nickvergessen commented on core/Controller/GenericAvatarController.php:
```diff
@@ -0,0 +1,173 @@
+			);
+		} catch (\Exception $e) {
+			$this->logger->logException($e, ['app' => 'core']);
+			return new JSONResponse(
```
```suggestion
			return new DataResponse(
```

nickvergessen commented on core/Controller/GenericAvatarController.php:
```diff
@@ -0,0 +1,173 @@
+		try {
+			$avatar = $this->avatarManager->getGenericAvatar($avatarType, $avatarId);
+			$avatar->remove();
+			return new JSONResponse();
```
```suggestion
			return new DataResponse();
```

nickvergessen commented on core/Controller/GenericAvatarController.php:
```diff
@@ -0,0 +1,173 @@
+	/**
+	 * @PublicPage
+	 *
+	 * @return JSONResponse
```
```suggestion
	 * @return DataResponse
```

nickvergessen commented on core/Controller/GenericAvatarController.php:
```diff
@@ -0,0 +1,173 @@
+			return new JSONResponse();
+		} catch (\Exception $e) {
+			$this->logger->logException($e, ['app' => 'core']);
+			return new JSONResponse(
```
```suggestion
			return new DataResponse(
```

danxuliu committed:
fixup! Add OCS endpoint for avatars

Signed-off-by: Daniel Calviño Sánchez <danxuliu@gmail.com>

danxuliu committed:
fixup! Add OCS endpoint for avatars

Signed-off-by: Daniel Calviño Sánchez <danxuliu@gmail.com>

danxuliu committed:
Make possible to define custom avatar types

Signed-off-by: Daniel Calviño Sánchez <danxuliu@gmail.com>

danxuliu committed:
fixup! Add OCS endpoint for avatars

Signed-off-by: Daniel Calviño Sánchez <danxuliu@gmail.com>

danxuliu committed:
Move registration of IAvatarProviders to IRegistrationContext

Signed-off-by: Daniel Calviño Sánchez <danxuliu@gmail.com>

ChristophWurst commented on core/Controller/GenericAvatarController.php:
```diff
@@ -0,0 +1,174 @@
+	 * @return DataResponse|FileDisplayResponse
+	 */
+	public function getAvatar(string $avatarType, string $avatarId, int $size): Response {
+		// min/max size
```
you can even use the `min`/`max` functions for this :wink:

nickvergessen replied:
but not for <= 0 use 64.

On the other side @rullzer  always wanted to limit the formats anyway.
I would suggest to only allow: 64, 128, 256 and 512 as sizes.

danxuliu replied:
Actually I have realized that this (which is a shameless copy/paste from [the old controller](https://github.com/nextcloud/server/blob/d9015a8c94bfd71fe484618a06d276701d3bf9ff/core/Controller/AvatarController.php#L117-L122) :-P ) does not allow to use -1 to get an avatar with the original size (which is [documented in `IAvatar`](https://github.com/nextcloud/server/blob/5bf3d1bb384da56adbf205752be8f840aac3b0c5/lib/public/IAvatar.php#L83)). Should we allow that in the new controller?

nickvergessen replied:
Only 64, 128, 256, 512 ?

danxuliu replied:
I have no real preference on the set of sizes, although limiting it to only a specific set sounds like a good idea to prevent an unneeded use of disk space if a "funny" user does something like `for i in {64..512}; do curl .../avatar/user0/$i; done`. Thus I have added a commit for that.

I still wonder about the "-1" to get the original size; it could be used for example if a user wants to check her current avatar exactly like it was uploaded.

ChristophWurst commented on core/Controller/GenericAvatarController.php:
```diff
@@ -0,0 +1,174 @@
+					'X-NC-IsCustomAvatar' => (int)$avatar->isCustomAvatar()
+				]
+			);
+		} catch (\Exception $e) {
```
* This is a bit generic – is it possible to catch a more specific exception?
* If catching that generic I would highly recommend logging the exception to help our futures selves debug any issues with this code :)

danxuliu replied:
> This is a bit generic – is it possible to catch a more specific exception?

We could catch `OCP\Files\NotFoundException` as it is thrown by [getFile()](https://github.com/nextcloud/server/blob/5bf3d1bb384da56adbf205752be8f840aac3b0c5/lib/public/IAvatar.php#L85), `InvalidArgumentException` as it is thrown by `IAvatarManager` and whatever is thrown by `IAvatarProvider::getAvatar()`. I wonder whether other exceptions should be catched or just left to bubble up and generate an error 500 (maybe even the exception thrown by `IAvatarProvider::getAvatar()` if it ends being just a generic `Exception`).

> If catching that generic I would highly recommend logging the exception to help our futures selves debug any issues with this code :)

:+1:

danxuliu replied:
Instead of `AvatarException` I named it `AvatarProviderException`, as that is more specific and gives room to add in the future an `AvatarException` for avatar specific errors if needed.

In the end I removed the log, as now a generic exception is no longer caught, but an `AvatarProviderException`, which would be thrown if the provider can not get an avatar for the given id. I think that belongs to the realm of expected errors that do not need to be logged (but if I am wrong it can be added back, of course ;-) ).

ChristophWurst commented on core/Controller/GenericAvatarController.php:
```diff
@@ -0,0 +1,174 @@
+				Http::STATUS_OK,
+				[
+					'Content-Type' => $avatarFile->getMimeType(),
+					'X-NC-IsCustomAvatar' => (int)$avatar->isCustomAvatar()
```
aren't headers always strings? cast to string or explicitly use the ternary operator to convert to `"0"` and `"1"`?

danxuliu replied:
```
php > echo (string)false;
php > echo (string)true;
1
```
Ternary operator is it then.

ChristophWurst commented on lib/private/Avatar/AvatarManager.php:
```diff
@@ -65,6 +68,15 @@ class AvatarManager implements IAvatarManager {
+	private $bootstrapCoordinator;
+
+	/** @var IAvatarProvider[] */
+	private $providers;
```
```suggestion
	private $providers = [];
```

can be used instead of the initialization in the constructor

ChristophWurst commented on lib/private/Avatar/AvatarManager.php:
```diff
@@ -152,4 +171,33 @@ public function deleteUserAvatar(string $userId): void {
+
+		if ($context === null) {
+			// Too early, nothing to do yet
+			throw new \InvalidArgumentException('Unknown avatar type: ' . $type);
```
I'd argue that we are lying here. The argument isn't invalid, but the code can't handle the request. Another exception would be more appropriate IMO.

Of course this is highly unlikely to happen because we do the registration before anything else. Hence failing hard here could be an option.

danxuliu replied:
> I'd argue that we are lying here. The argument isn't invalid, but the code can't handle the request. Another exception would be more appropriate IMO.

I agree, although I used `InvalidArgumentException` because I did not find any suitable exception and adding a new exception just for this seemed unnecessary (given that in practice it will not happen). So exception suggestions welcome :-)

> Of course this is highly unlikely to happen because we do the registration before anything else. Hence failing hard here could be an option.

Failing hard as in `exit()`? Having an exit there would feel very out of place in my opinion. Or do you mean `throw new Error()`? Or something else? :-)

ChristophWurst replied:
> I agree, although I used `InvalidArgumentException` because I did not find any suitable exception and adding a new exception just for this seemed unnecessary (given that in practice it will not happen). So exception suggestions welcome :-)

In https://github.com/nextcloud/server/pull/24702/files#diff-7bfcce99ae84563547086798084d75003c7099fedeabb795d2c01962293b33fcR88 I went with `\RuntimeException`.

> Failing hard as in `exit()`? Having an exit there would feel very out of place in my opinion. Or do you mean `throw new Error()`? Or something else? :-)

I meant letting the exception bubble up, without any explicit handling :)

danxuliu replied:
> In https://github.com/nextcloud/server/pull/24702/files#diff-7bfcce99ae84563547086798084d75003c7099fedeabb795d2c01962293b33fcR88 I went with \RuntimeException.

`\RuntimeException` it is then.

> > Failing hard as in exit()? Having an exit there would feel very out of place in my opinion. Or do you mean throw new Error()? Or something else? :-)
>
> I meant letting the exception bubble up, without any explicit handling :)

Ok, that makes a lot more sense :-P

ChristophWurst commented on lib/public/IAvatarManager.php:
```diff
@@ -56,4 +56,17 @@ public function getAvatar(string $user) : IAvatar;
+	 * @return IAvatar
+	 * @throws \InvalidArgumentException if the type is not known or the id is
+	 *         not valid
+	 * @throws \Exception if getting the avatar failed
```
I'd like to see a more specific exception thrown here. If this is just about *any possible error that we didn't expect* then I wouldn't document the exception

danxuliu replied:
It should be the same exception thrown by `IAvatarProvider::getAvatar()`. Should `AvatarException` be introduced to wrap other specific exceptions that could be thrown by providers? For example, in the case of Talk, the avatar provider could throw `RoomNotFoundException` if the avatar id does not match any room token, but of course that exception makes little sense in a generic context.

ChristophWurst replied:
> Should `AvatarException` be introduced to wrap other specific exceptions

Sounds good to me! Then we can transport any more specific exception as `previous`.

ChristophWurst commented on lib/public/IAvatarProvider.php:
```diff
@@ -0,0 +1,46 @@
+interface IAvatarProvider {
+
+	/**
+	 * Returns an IAvatar instance for the given id.
```
no `.` at the end of phpdoc lines :)

danxuliu replied:
I will remove all them, except at the end of `It is allowed to register more than one provider per app.` (`IRegistrationContext::registerAvatarProvider`), as:
- I guess the rule does not apply to the "body" of a documentation block (but if it does please let me know ;-) )
- it is consistent with the rest of the documentation in `IRegistrationContext`

But if that should be removed too let me know ;-)

ChristophWurst replied:
There is no hard rule, I guess. For the body it seems fine. The first line (title?) is a bit strange.

But this is just nitpicking. You may also just ignore my remark ;)

danxuliu replied:
> You may also just ignore my remark ;)

Too late ;-)

ChristophWurst reviewed (changes requested):
I like this a lot because we can also use this to unify/replace https://github.com/nextcloud/mail/blob/master/lib/Controller/AvatarsController.php

ChristophWurst commented:
> Should it be done in `UserAvatar`? Or should be prevented only in the controller (for example, to allow an admin to modify the avatar from an OCC command, or something like that)? In case it needs to be done in the controller, how? Add something like `canAvatarBeModifiedByUser($userId)` to `IAvatar` and check it from the controller?

We could also make this part of the new avatar provider, right? And then the controller asks the avatar service, which delegates to the provider (or falls back to `false` or whatever)

But now I'm also thinking about whether a avatar is just identified by `(type, id)` or if there should be an optional third param for `uid`. Then we only let users update *their* avatars? Or the provider internally checks whether the "current" user is allowed to change that avatar? Like if a user updates an avatar of a talk conversation, the Talk provider could know if that UID is moderator and only then it tells the service/provider that modification is fine.

Alternatively we don't even ask for permission but just let the providers throw a dedicated exception when the operation isn't allowed. But again we will likely need the UID as context.

> Where to specify the cache value to use for each type? With "getCache()" in the AvatarProvider? With a more generic "getAvatarOptions()" (or something similar) that returns a map of keys and values so IAvatarProvider does not need to be extended? In the IAvatar itself so it could even change depending on the avatar instance?

A ``IAvatar::getAvatarOptions`` could give us lots of flexibility with the price of no type guarantees. Also if this interface is implemented by any app, the app will break with this change.

Therefore I'm more in favor of the ``IAvatarProvider::getCache``. Maybe like a ``IAvatarProvider::getCacheTtl(IAvatar): ?int`` so the provider can have the *depending on the avatar instance* flexibility or ignore the parameter. Returning `null` could signal that there shouldn't be any caching.

danxuliu commented:
> We could also make this part of the new avatar provider, right? And then the controller asks the avatar service, which delegates to the provider (or falls back to false or whatever)
> 
> But now I'm also thinking about whether a avatar is just identified by (type, id) or if there should be an optional third param for uid. Then we only let users update their avatars? Or the provider internally checks whether the "current" user is allowed to change that avatar? Like if a user updates an avatar of a talk conversation, the Talk provider could know if that UID is moderator and only then it tells the service/provider that modification is fine.
> 
> Alternatively we don't even ask for permission but just let the providers throw a dedicated exception when the operation isn't allowed. But again we will likely need the UID as context.

Wouldn't it be possible to inject the current user in the avatar provider constructor? Or that is possible only for controllers? In that case the extra parameter would be unneeded, as the IAvatarProvider subclass would define everything it needs to know if the current user can modify (or even get) an avatar or not. Even if the current user is not a real user, but for example a guest (in Talk currently only moderators would be able to set the avatar of a conversation, but technically that could be possible also for a guest moderator).

> Therefore I'm more in favor of the IAvatarProvider::getCache. Maybe like a IAvatarProvider::getCacheTtl(IAvatar): ?int so the provider can have the depending on the avatar instance flexibility or ignore the parameter. Returning null could signal that there shouldn't be any caching.

Fully agree, except maybe in the name. Personally I prefer to avoid abbreviatures in API names (or in code names in general), so I would call it `getCacheTimeToLive(IAvatar)` instead.

danxuliu committed:
fixup! Make possible to define custom avatar types

Signed-off-by: Daniel Calviño Sánchez <danxuliu@gmail.com>

danxuliu committed:
fixup! Add OCS endpoint for avatars

Signed-off-by: Daniel Calviño Sánchez <danxuliu@gmail.com>

danxuliu committed:
fixup! Add getter for generic avatars to IAvatarManager

Signed-off-by: Daniel Calviño Sánchez <danxuliu@gmail.com>

danxuliu committed:
fixup! Make possible to define custom avatar types

Signed-off-by: Daniel Calviño Sánchez <danxuliu@gmail.com>

danxuliu committed:
fixup! Add OCS endpoint for avatars

Signed-off-by: Daniel Calviño Sánchez <danxuliu@gmail.com>

danxuliu committed:
Move deprecated ILogger to LoggerInterface in avatar private classes

Signed-off-by: Daniel Calviño Sánchez <danxuliu@gmail.com>

danxuliu committed:
fixup! Move registration of IAvatarProviders to IRegistrationContext

Signed-off-by: Daniel Calviño Sánchez <danxuliu@gmail.com>

danxuliu committed:
Add explicit providers for user and guest avatars

User and guest avatars were treated as built-in types and handled
directly by the AvatarManager. Now a provider is introduced for each
type, so they are handled like any generic avatar type.

The AvatarManagerTest only tested getting user avatars, so the unit test
was renamed and adjusted to test the UserAvatarProvider instead.

Signed-off-by: Daniel Calviño Sánchez <danxuliu@gmail.com>

danxuliu committed:
Remove no longer needed attributes from AvatarManager

Signed-off-by: Daniel Calviño Sánchez <danxuliu@gmail.com>

danxuliu committed:
fixup! Make possible to define custom avatar types

Signed-off-by: Daniel Calviño Sánchez <danxuliu@gmail.com>

danxuliu committed:
fixup! Make possible to define custom avatar types

Signed-off-by: Daniel Calviño Sánchez <danxuliu@gmail.com>

danxuliu committed:
fixup! Move registration of IAvatarProviders to IRegistrationContext

Signed-off-by: Daniel Calviño Sánchez <danxuliu@gmail.com>

danxuliu committed:
Get the avatar provider rather than the avatar itself from the manager

This will make possible to use other future IAvatarProvider methods.

Signed-off-by: Daniel Calviño Sánchez <danxuliu@gmail.com>

danxuliu committed:
Add method to get the cache duration of an avatar to IAvatarProvider

Signed-off-by: Daniel Calviño Sánchez <danxuliu@gmail.com>

danxuliu commented:
I did something wrong in `Add explicit providers for user and guest avatars`, as `appData` is not properly resolved, but I can not find what. Hopefully I will be more inspired tomorrow :shrug:

danxuliu committed:
fixup! Add method to get the cache duration of an avatar to IAvatarProvider

Signed-off-by: Daniel Calviño Sánchez <danxuliu@gmail.com>

nickvergessen commented on lib/public/IAvatarManager.php:
```diff
@@ -58,15 +58,12 @@ public function getAvatar(string $user) : IAvatar;
-	 * @throws \Exception if getting the avatar failed
+	 * @return IAvatarProvider
+	 * @throws \InvalidArgumentException if the type is not known
 	 * @since 21.0.0
```
🙈 quick quick quick

ChristophWurst commented:
> Wouldn't it be possible to inject the current user in the avatar provider constructor?

Yes, that should be possible. And it will also work just fine. Personally I sometimes prefer having the user as method argument because then the method (and the service it belongs to) becomes more independent and can be used in background processes, e.g. to iterate over a bunch of users and calling that method for each of them. You can't do this with a constructor arg, unless you create a new instance of the class for every user. So judge for yourself on what is better here. I'm also fine with the constructor injected user ID :)

> would call it `getCacheTimeToLive(IAvatar)` instead

Sounds good!

danxuliu committed:
fixup! Add explicit providers for user and guest avatars

Signed-off-by: Daniel Calviño Sánchez <danxuliu@gmail.com>

danxuliu committed:
fixup! Move registration of IAvatarProviders to IRegistrationContext

Signed-off-by: Daniel Calviño Sánchez <danxuliu@gmail.com>

danxuliu committed:
fixup! Get the avatar provider rather than the avatar itself from the manager

Signed-off-by: Daniel Calviño Sánchez <danxuliu@gmail.com>

danxuliu committed:
fixup! Make possible to define custom avatar types

Signed-off-by: Daniel Calviño Sánchez <danxuliu@gmail.com>

danxuliu committed:
fixup! Add explicit providers for user and guest avatars

Signed-off-by: Daniel Calviño Sánchez <danxuliu@gmail.com>

danxuliu committed:
fixup! Add explicit providers for user and guest avatars

Signed-off-by: Daniel Calviño Sánchez <danxuliu@gmail.com>

danxuliu committed:
fixup! Get the avatar provider rather than the avatar itself from the manager

Signed-off-by: Daniel Calviño Sánchez <danxuliu@gmail.com>

danxuliu committed:
fixup! Add integration tests for getting and setting generic avatars

Signed-off-by: Daniel Calviño Sánchez <danxuliu@gmail.com>

danxuliu committed:
fixup! Add integration tests for getting and setting generic avatars

Signed-off-by: Daniel Calviño Sánchez <danxuliu@gmail.com>

danxuliu committed:
fixup! Add OCS endpoint for avatars

Signed-off-by: Daniel Calviño Sánchez <danxuliu@gmail.com>

danxuliu committed:
fixup! Make possible to define custom avatar types

Signed-off-by: Daniel Calviño Sánchez <danxuliu@gmail.com>

danxuliu committed:
fixup! Add integration tests for getting and setting generic avatars

Signed-off-by: Daniel Calviño Sánchez <danxuliu@gmail.com>

danxuliu committed:
fixup! Add method to get the cache duration of an avatar to IAvatarProvider

Signed-off-by: Daniel Calviño Sánchez <danxuliu@gmail.com>

danxuliu committed:
Split try/catch blocks to catch only the relevant exceptions

Signed-off-by: Daniel Calviño Sánchez <danxuliu@gmail.com>

danxuliu committed:
Add method to check if an avatar can be accessed by the current user

IAvatarProvider implementers now need to define whether the current user
can access or not a given IAvatar (created by the provider itself). With
this the GenericAvatarController can allow or deny access to avatars in
a generic way.

The controller returns STATUS_NOT_FOUND instead of STATUS_FORBIDDEN to
prevent revealing to unauthorized users the existence of an element with
the given id.

Adding the method to IAvatarProvider prevents touching IAvatar, which
might require changes in apps, and also keeps the access control logic
separate from the avatar logic, which is more focused on handling of
files and images.

Signed-off-by: Daniel Calviño Sánchez <danxuliu@gmail.com>

danxuliu committed:
Add method to check if an avatar can be modified by the current user

IAvatarProvider implementers now need to define whether the current user
can modify or not a given IAvatar (created by the provider itself). With
this the GenericAvatarController can allow or deny modification
(including deletion) of avatars in a generic way.

The controller returns STATUS_NOT_FOUND instead of STATUS_FORBIDDEN to
prevent revealing to unauthorized users the existence of an element with
the given id.

Adding the method to IAvatarProvider prevents touching IAvatar, which
might require changes in apps, and also keeps the access control logic
separate from the avatar logic, which is more focused on handling of
files and images.

Signed-off-by: Daniel Calviño Sánchez <danxuliu@gmail.com>

danxuliu committed:
Add integration tests for unauthorized modification of avatars

Signed-off-by: Daniel Calviño Sánchez <danxuliu@gmail.com>

danxuliu committed:
Add method to get the version of an avatar

Signed-off-by: Daniel Calviño Sánchez <danxuliu@gmail.com>

danxuliu committed:
Limit the returned sizes of the avatars to a predefined set

Signed-off-by: Daniel Calviño Sánchez <danxuliu@gmail.com>

danxuliu commented:
> Yes, that should be possible. And it will also work just fine. Personally I sometimes prefer having the user as method argument because then the method (and the service it belongs to) becomes more independent and can be used in background processes, e.g. to iterate over a bunch of users and calling that method for each of them. You can't do this with a constructor arg, unless you create a new instance of the class for every user. So judge for yourself on what is better here. I'm also fine with the constructor injected user ID :)

In this case, which is checking whether the current user can get or modify an avatar or not, I do not think that there would be a need to call the method with different users, so injecting it in the constructor of the provider should be enough.

On a different matter, I have added a `getVersion(IAvatar)` method to the `IAvatarProvider` as I think that it may come handy to generate the `?v=XXX` when requesting avatars (I have not touched any of the code in the server that does that, although I expect to use it in Talk).

Regarding being able to [return an URL besides just the actual image](https://github.com/nextcloud/mail/blob/0a1e1ad8bc46bb4d4b0846b840cd6b8a4ff3c408/lib/Controller/AvatarsController.php#L61) so all this can be used from Mail too I would defer that for later and just settle with the current API for an initial version (which, even if it will need to be maintained, I would say that it is a _reasonable enough_ API).

I think that everything else is addressed, but let me know if I forgot something (or if you think that adding a way to get the URL for an avatar is a must ;-) ). If not I will do the rebase and (finally!) set it as ready for review.

danxuliu committed:
fixup! Get the avatar provider rather than the avatar itself from the manager

Signed-off-by: Daniel Calviño Sánchez <danxuliu@gmail.com>

danxuliu committed:
fixup! Add explicit providers for user and guest avatars

Signed-off-by: Daniel Calviño Sánchez <danxuliu@gmail.com>

danxuliu committed:
fixup! Add method to check if an avatar can be modified by the current user

Signed-off-by: Daniel Calviño Sánchez <danxuliu@gmail.com>

danxuliu committed:
fixup! Add explicit providers for user and guest avatars

Signed-off-by: Daniel Calviño Sánchez <danxuliu@gmail.com>

danxuliu committed:
fixup! Make possible to define custom avatar types

Signed-off-by: Daniel Calviño Sánchez <danxuliu@gmail.com>

ChristophWurst commented:
I'll try to quickly adapt Mail for this to get a better understanding. But I think it's fine.

One more thing I noticed is that the public API is in the ``\OCP`` root. We might want to have this in a subdirectory/subnamespace.

nickvergessen commented on core/Controller/GenericAvatarController.php:
```diff
@@ -0,0 +1,242 @@
+			}
+		}
+
+		return $size;
```
Could also just have checked if its in valid size and otherwise use 64 or return 404.
But now you implemented it already 🙃

ChristophWurst commented:
> In this case, which is checking whether the current user can get or modify an avatar or not, I do not think that there would be a need to call the method with different users, so injecting it in the constructor of the provider should be enough.

Now that I'm implementing my own provider I feel even stronger about making the user ID (or even the user object) an argument of ``\OCA\Mail\Avatar\AvatarProvider::canBeAccessedByCurrentUser`` and ``\OCA\Mail\Avatar\AvatarProvider::canBeModifiedByCurrentUser``. In my perception the API is simpler if the information you need to carry out the operation is provided by the caller. It's not just about the ability to call the method in a context without a *current user*. It really feels like avoidable extra work to have the UID injected in each of the providers. It would make the API also a bit more *functional* and thus easier to reason about. Testing with different UIDs also becomes very easy.

Another remark: will we offer any helpers for ``\OCA\Mail\Avatar\AvatarProvider::getAvatar``? The current ``\OCP\IAvatar`` is rather bloated. Looks at this:

![Bildschirmfoto von 2020-12-16 08-12-21](https://user-images.githubusercontent.com/1374172/102316566-72399800-3f76-11eb-8198-6c883f1226c3.png)

I have to implement all those for my custom Mail avatars. 

* ``exists`` doesn't make sense – why would I provide an ``IAvatar`` that doesn't exist
* ``isCustomAvatar`` will always be ``true`` – there is no need for any app developer to define this
* ``set`` only makes sense for user-provided avatars – in Mail we don't need this – maybe there should be a dedicated interface for *mutable avatars*
* ``remove`` see ``set``
* ``getFile`` makes me realize that the current avatars have a strong assumption about all avatar images being local files. In the case of Mail we use favicons, gravatar and contact images as source. We don't ever store those. It would just bloat the app data with manual cache data. I'd like to avoid this. We should make it possible to also have only a stream to a file. After all what we do is return a HTTP response with the image. We don't need a physical file on storage for this.
* ``avatarBackgroundColor`` possibly a method that never should have been added to the public interface, but unrelated to this discussion. But I highly doubt that every app that provides an avatar type should define their own logic for the avatarBackgroundColor. This logic should be internal to Nextcloud when no (custom) avatar can be provided.
* ``userChanged`` a method that should have been an event, I guess. We should deprecated this. But as for the ``set`` remarks, only mutable avatars would be interested in this. And even then it's about a *changed user*, so really just relevant for *user avatar* and no custom avatar.

All this leads me to think that it's maybe better if we start a new, simplified base interface for avatars with only those aspects that every generic avatar needs. User-specific aspects should remain part of the server's user avatars. Avatar mutation could be a sub-interface.

Also I see that the ``\OCP\IAvatarProvider`` has some kind of generic aspect in its types. Like ``getAvatar`` returns *some* implementation of ``IAvatar``, and the other four methods take the same base interface type as argument. Are they supposed to handle *any* ``IAvatar`` or just their own? Php won't let us express generics. But we can tell Psalm that the ``IAvatarProvider`` is generic with a template parameter for the returned ``IAvatar``, which the other methods take as input. Then the implementers can make some assumption about the data they receive. But those semantics also have to be implemented in our code.

Summing up I think that the current design is doable, but with tradeoffs and *known* technical dept. Technical dept that we'll have to carry for three years (~9 major releases).

> while working on the [_Group conversation picture_ feature for Talk](https://github.com/nextcloud/spreed/issues/927)

Is this feature a must-have for 21? As in, are we blocking your work when we say that we delays custom avatar types for 22?

I'm sorry for the late blocking. But I'm feeling that we rushing this and that we need a bit more time to settle with a good design that we can use the next few years.

nickvergessen commented on lib/public/AvatarProviderException.php:
```diff
@@ -0,0 +1,44 @@
+	 * @param int $code
+	 * @param \Exception $previous
+	 */
+	public function __construct(string $message = "", int $code = 0, \Exception $previous = null) {
```
```suggestion
	public function __construct(string $message = '', int $code = 0, \Exception $previous = null) {
```

ChristophWurst replied:
and for ``$previous`` we should allow ``\Throwable``, not just ``\Exception``

nickvergessen commented on lib/public/IAvatarProvider.php:
```diff
@@ -0,0 +1,113 @@
+	 *         this provider
+	 * @since 21.0.0
+	 */
+	public function canBeAccessedByCurrentUser(IAvatar $avatar): bool;
```
I agree with christoph that we should inject the user here

nickvergessen commented on lib/public/IAvatarProvider.php:
```diff
@@ -0,0 +1,113 @@
+	 *         this provider
+	 * @since 21.0.0
+	 */
+	public function canBeModifiedByCurrentUser(IAvatar $avatar): bool;
```
I agree with @ChristophWurst  that we should inject the user here

nickvergessen commented on lib/public/IAvatarProvider.php:
```diff
@@ -0,0 +1,113 @@
+	public function canBeModifiedByCurrentUser(IAvatar $avatar): bool;
+
+	/**
+	 * Returns the latest value of the avatar version
+	 *
+	 * Implementers of IAvatarProvider may not throw \InvalidArgumentException
+	 * if the behaviour does not depend on specific avatar instances (for
+	 * example, if versions are not supported and the same version is always
+	 * returned).
+	 *
```
Now you described what it does. But what is it used for? Caching of the avatars? Do we need to bump it on each patch we do?

ChristophWurst commented:
If this is a must-have for 21 then I would suggest that we make this an internal API only. So Talk can use it it its current form. For 22 we brainstorm together and find something more suitable for the general purpose and that can be our public API, which Talk uses from then on.

That means you can continue the feature and we don't have to suffer with the tech dept for three years :relieved:

nickvergessen commented:
Alternatively we could just implement the route:
```
['root' => '/core', 'name' => 'GenericAvatar#getAvatar', 'url' => '/avatar/talk/{avatarId}/{size}', 'verb' => 'GET'],
```
In Talk for 21 directly. So it will work for the clients already and when the API is merged and implemented by Talk for 22, we remove that route take over and it uses the GenericAvatar thing from core

danxuliu commented:
> Alternatively we could just implement the route in Talk for 21 directly

That was my fallback plan, so I would do that rather than adding this as an internal API.

MorrisJobke commented:
@danxuliu @nickvergessen What is the status here?

nickvergessen commented:
Move to 23 once the milestone exists

MorrisJobke commented:
> Move to 23 once the milestone exists

Done.
